docs/test_overview.md

```diff
@@ -0,0 +1,249 @@
+
+```cpp
+// 0) Define test name and tags
+CUB_TEST("SCOPE FACILITY works with CONDITION",
+         "[FACILITY][SCOPE]") 
+{
+  using type = std::int32_t;
+  constexpr int threads_per_block = 256;
+  constexpr int num_items = threads_per_block;
+
+  // 1) Allocate device input
+  thrust::device_vector<type> d_input(num_items);
+
+  // 2) Generage 3 random input arrays using Catch2 helper
+  c2h::gen(CUB_SEED(3), d_input);
+
+  // 3) Allocate output array
+  thrust::device_vector<type> d_output(d_input.size());
+
+  // 4) Copy device input to host
+  thrust::host_vector<key_t> h_reference = d_input;
+
+  // 5) Compute reference output
+  std::ALGORITHM(
+      thrust::raw_pointer_cast(h_reference.data()), 
+      thrust::raw_pointer_cast(h_reference.data()) + h_reference.size());
+
+  // 6) Comute CUB output
+  SCOPE_ALGORITHM<threads_per_block>(d_input.data(),
+                                     d_output.data(),
+                                     d_input.size());
+
+  // 7) Compare device and host results
+  REQUIRE( d_input == d_output );
+}
+```
+
+We introduce test cases with the `CUB_TEST` macro in (0). 
+This macro always takes two string arguments - a free-form test name and
+one or more tags. Then, in (1), we allocate device memory using `thrust::device_vector`.
+The memory is filled with random data in (2).
+
+Generator `c2h::gen` takes two parameters. 
+The first one is a random generator seed. 
+Instead of providing a single value, we use the `CUB_SEED` macro.
+The macro expects a number of seeds that has to be generated. 
+In the example above, we require three random seeds to be generated.
+This leads to the whole test being executed three times 
+with different seed values. 
+
+Later, 
+in (3) and (4),
+we allocate device output and host reference.
+Then, in (4), 
+we populate host input data and perform reference computation on the host in (5).
+Then launch the CUB algorithm in (6). 
+At this point, we have a reference solution on CPU and CUB solution on GPU.
+The two can be compared with `REQUIRE` assert. 
+
+> :warning: Standard algorithms (`std::`) have to be used as much as possible when computing reference solutions.
+
+If your test has to cover floating point types, 
+it's sufficient to replace `REQUIRE( a == b )` with `REQUIRE_APPROX_EQ(a, b)`.
+
+It's strongly advised to always use `c2h::gen` to produce input data.
+Other data generation methods might be used 
+if absolutely necessary in tests of corner cases.
+
+If a custom type has to be tested, the following helper should be used:
+
+```cpp
+using type = c2h::custom_type_t<c2h::accumulateable_t,
+                                c2h::equal_comparable_t>;
+```
+
+Here we enumerate all the type properties that we are interested in.
+The produced type ends up having `operator==` and `operator+`.
+There are more properties implemented. 
+If some property is missing, 
+it'd be better to add one in `c2h` 
+instead of writing a custom type from scratch.
```

Comment:
Wondering if we should give this part its own section or fuse it with _Type Lists_